### PR TITLE
Added Main Readme and env file for build

### PR DESCRIPTION
**Pull request by mvvsmk**
I have described the project in brief and have given instructions about the build and also added one onnx command to invoke each ML driven optimization. I have also added the env file required for the build in the home directory.

### CODE REVIEW

**Comment by mvvsmk** [comment]
There is one issue that is not resolved and that is of the `ml-config-folder` people are still copying it from Pramana , how would you like me to incorporate it into the readme

File: README.md [hunk 1]
```diff
@@ -1,271 +1,215 @@
...
-<LABEL_ID> [label=" {RegClass} {SpillCost} <Matrix of Live Range/Interval>"];
-<LABEL_ID>--<LABEL_ID>; // Show the interference between two Live Range/Interval
-}
-```
+# ML LLVM Project
+
+## Contents
+-	[About](#about)
+-	[Setup](#setup)
+    - Requirements
```
Comment (svkeerthy) [inline README.md]: Please add links to each step

File: README.md [hunk 2]
```diff
@@ -1,271 +1,215 @@
...
+
+## About
+This GitHub repository encompasses the complete Compiler Infrastructure for ML-Driven Optimizations developed by researchers at IIT H. The repository serves as a valuable reference, illustrating the integration of ML-driven optimization techniques into the LLVM project through the ML Compiler Bridge infrastructure.
+
+We strongly encourage you to delve into this repository, explore its contents, and consider building additional tools leveraging the existing infrastructure. We presume you are fimiliar with LLVM and build upon that, but if you are not fimiliar with llvm them, here are a few resources that might help :
+
+* [Getting Started with LLVM](https://llvm.org/docs/GettingStarted.html#getting-started-with-llvm)
+page for detailed information on configuring and compiling LLVM. You can visit
+* [Directory Layout](https://llvm.org/docs/GettingStarted.html#directory-layout)
+to learn about the layout of the source code tree.
```
Comment (svkeerthy) [inline README.md:29]: Please add about IR2Vec here

File: README.md [hunk 3]
```diff
@@ -1,271 +1,215 @@
...
+As a part of the ML-Compiler-Bridge [research](https://arxiv.org/pdf/2311.10800.pdf), it is possible to have multiple ways of integrating compiler and the Machine learning model. These methods primarily use server client communication techniques like gRPC, and pipes. The ONNX flow which is capable of representation of ML models into DAG-based IRs with callable APIs in multiple langugages (C/C++/Python),doesnt require a server-client model or inter process communication.
+> The Next 700 ML-Enabled Compiler Optimizations: S.VenkataKeerthy, Siddharth Jain, Umesh Kalvakuntla, Pranav Sai Gorantla, Rajiv Sailesh Chitale, Eugene Brevdo, Albert Cohen, Mircea Troffin, Ramakrishna Upadrasta
+
+## Setup
+
+### Requirements
+
+* cmake (>= 3.10)
+* GNU Make (4.2.1)
+* LLVM (10.X) - [src](https://github.com/llvm/llvm-project/tree/release/10.x), [release](https://releases.llvm.org/download.html#10.0.1) ## ask isn't it included with the repo
```
Comment (svkeerthy) [inline README.md]: This is not a requirement.

File: README.md [hunk 4]
```diff
@@ -1,271 +1,215 @@
...
-            "mapping": {
-                "1": 20
-            }
-        }
-    ]
-}
+* [ONNXRuntime](https://github.com/microsoft/onnxruntime/releases) v1.16.3
+    
+    * The following commands will download ONNX Runtime v1.16.3 in your present working directory and then untar the contents.
+    The path for this will be used in this [section](#exporting-onnx-path-variables)
```
Comment (svkeerthy) [inline README.md:66]: Recheck path.

File: README.md [hunk 5]
```diff
@@ -1,271 +1,215 @@
...
-
-### Training
-
-Add path to directory containing LLVM IR files to be used for training in [RLLib-PhaseOrder/Environment.py](RLLib-PhaseOrder/Environment.py)
-
-`python experiment.py --llvm_dir <Path to llvm build directory> --ir2vec_dir <Path to directory with IR2Vec binary and seed embedding>`
-
-### Inference
+> [!NOTE]
+>TODO: mlbridge.yml should be replaced with a unanimous env
```
Comment (svkeerthy) [inline README.md]: What is this?

File: README.md [hunk 6]
```diff
@@ -1,271 +1,215 @@
...
+    
+    * The following commands will download ONNX Runtime v1.16.3 in your present working directory and then untar the contents.
+    The path for this will be used in this [section](#exporting-onnx-path-variables)
+```bash
+	 wget https://github.com/microsoft/onnxruntime/releases/download/v1.16.3/onnxruntime-linux-x64-1.16.3.tgz
+	 tar -xvf onnxruntime-linux-x64-1.16.3.tgz
 ```
+* TensorFlow - for TF Model Runner (AOT flow)
+    * Tested with TensorFlow 2.13.0
+* Other python requirements are available in [mlbridge.yml]
```
Comment (svkeerthy) [inline README.md]: Broken path.

File: README.md [hunk 7]
```diff
@@ -1,271 +1,215 @@
...
+ export PATH=${ONNX_DIR}/include:$PATH 
+```
+> [!TIP] 
+> It is adviced to add these commands to your **~/.bashrc** as the'll be needed when you switch shells.
 
+#### Conda environment set-up
+The following commands will help you install the and set up the nessesary conda environments.
+```bash
+# install the env using the following commands
+conda env create -f ./mlgo-new.yml
```
Comment (svkeerthy) [inline README.md]: Can we rename the file and related things to `mlopt`?

File: README.md [hunk 8]
```diff
@@ -1,271 +1,215 @@
...
-`conda env create -f poset-rl-odg.yml`
+```bash
+# create a build dir and move to it
+mkdir build
+cd build
+```
+After moving to the build directory, we'll use CMake to generate our build files and directories
+
+```bash
+cmake -G "Unix Makefiles" -S ../llvm -B . \
```
Comment (svkeerthy) [inline README.md:130]: This need not have INLINER_MODEL and RAEVICT_MODEL related paths. Please remove and test. Then make changes.

Comment (svkeerthy) [inline README.md:130]: This need not be Makefile. It could be any generator that LLVM supports.

File: README.md [hunk 9]
```diff
@@ -1,271 +1,215 @@
...
+	-DLLVM_INLINER_MODEL_CURRENT_URL=https://github.com/google/ml-compiler-opt/releases/download/inlining-Oz-v1.1/inlining-Oz-99f0063-v1.1.tar.gz \
+	-DLLVM_RAEVICT_MODEL_PATH=download \
+	-DLLVM_RAEVICT_MODEL_CURRENT_URL=https://github.com/google/ml-compiler-opt/releases/download/regalloc-evict-v1.0/regalloc-evict-e67430c-v1.0.tar.gz	       
+```
 
-`conda activate poset-rl-odg`
+#### Make command
+After following the above steps you have successfully exproted all the required environment variables and have also created the Makefile which shall be used to build the project. Use the following command to start your build.
+```bash
+make clang opt -j50
```
Comment (svkeerthy) [inline README.md]: Make it generic please. make, j50 could be written better.

File: README.md [hunk 10]
```diff
@@ -1,271 +1,215 @@
...
+```
+> [!WARNING]  
+> Do not make all, your build will break. Only make clang and opt
```
Comment (svkeerthy) [inline README.md]: You can say make all is broken now and we will be fixing it.

File: README.md [hunk 11]
```diff
@@ -1,271 +1,215 @@
...
+```bash
+# ONNX command for inference:
+# this script will generate the optimized llfile
+./build/bin/opt -S \
+	-custom_loop_distribution \
+	-cld-use-onnx \
+	-ml-config-path= # path to your ml config  \
+	<file name> 
+```
+to learn more head to the Pass specific readme [here].
```
Comment (svkeerthy) [inline README.md]: broken link

File: README.md [hunk 12]
```diff
@@ -1,271 +1,215 @@
...
 
-`bash run-inference.sh`
+#### Try it out
+```bash
+./build/bin/clang  -O3 -mllvm \ #use clang or clang++ depending on your file type
+  -regalloc=greedy -mllvm -mlra-inference -mllvm \
+  -ml-config-path= #path to your ml config  \ 
+  -mllvm -rl-inference-engine \
+  <input_data_file>
+```
```
Comment (svkeerthy) [inline README.md:199]: Please add path to the pass.

File: README.md [hunk 13]
```diff
@@ -1,271 +1,215 @@
...
 
-Print size, throughput and sub-sequences chosen by the model to a csv
+### POSET-RL
 
-`bash results-binsize-reuse`
+POSET-RL uses a reinforcement learning approach as the search space of optimization sequences is too big to enumerate. For a compiler with m optimization passes, if the sequence length is fixed as n, then there can be potentially mn combinations, allowing repetitions. The reinforcement learning model is trained and evaluated on programs that are represented using IR2Vec embeddings. The action space contains the subsequences created using the Oz dependence graph (ODG). Sequences are constructed from this graph by finding walks that start and end at critical nodes (with degree greater than a value k).[slides](https://llvm.org/devmtg/2022-04-03/slides/POSET-RL.Phase.ordering.for.Optimizing.Size.and.Execution.Time.using.Reinforcement.Learning.pdf)
```
Comment (svkeerthy) [inline README.md]: Not much details are required. Please reduce.

Comment (svkeerthy) [reply README.md]: Add links to the paper, webpage, etc.

Comment (svkeerthy) [reply README.md]: Also point to the pass.

File: README.md [hunk 14]
```diff
@@ -1,271 +1,215 @@
...
 
-Clean temporary files generated
+> POSET-RL: Phase ordering for Optimizing Size and Execution Time using Reinforcement Learning: Shalini Jain, Yashas Andaluri, S. VenkataKeerthy and Ramakrishna Upadrasta
```
Comment (svkeerthy) [inline README.md]: Please add conference name and year. It should be ISPASS 2022.

File: README.md [hunk 15]
```diff
@@ -1,271 +1,215 @@
...
 
-`python experiment.py --llvm_dir <Path to llvm build directory> --ir2vec_dir <Path to directory with IR2Vec binary and seed embedding>`
+### RL4Real
 
-### Inference
+`RL4ReAl` is a retargetable Reinforcement Learning (RL) approach for solving the REgister ALlocation (REAL) problem on diverse architectures.
 
-Add paths to `llvm_dir`, `ir2vec_dir` and saved RLLib model to run-inference.sh
+This is described in the paper [here](https://dl.acm.org/doi/abs/10.1145/3578360.3580273) ([arXiv](https://arxiv.org/abs/2204.02013)).
+Please see [here](https://compilers.cse.iith.ac.in/publications/rl4real/) for more details.
```
Comment (svkeerthy) [inline README.md:189]: Please add reference to the paper like what you have in distribution and poset-rl

File: README.md [hunk 16]
```diff
@@ -1,271 +1,215 @@
...
-## Experiments
-Install and activate the conda environment
+We propose a Reinforcement Learning (RL) approach for loop distribution, optimizing for both vectorization and locality. Using SCC Dependence Graphs (SDGs), our RL model learns loop distribution order through topological walks. The reward is based on instruction cost and cache misses. We introduce a strategy to expand the training set by generating new loops. This method aims to enhance loop parallelization and improve overall code performance.
 
-`conda env create -f rllib_env.yml`
+This is described in the paper [here](https://ieeexplore.ieee.org/abstract/document/10026979) .
+Please see [here](https://compilers.cse.iith.ac.in/publications/rl_loop_distribution/) for more details.
 
-`conda activate rllib_env`
+> Reinforcement Learning assisted Loop Distribution for Locality and Vectorization, Shalini Jain, S. VenkataKeerthy, Rohit Aggarwal, Tharun Kumar Dangeti, Dibyendu Das, Ramakrishna Upadrasta
```
Comment (svkeerthy) [inline README.md]: Add conference venue and year. It should be LLVM-HPC, 2022.

**Review by svkeerthy — CHANGES_REQUESTED** [review]

File: README.md [hunk 17]
```diff
@@ -1,271 +1,216 @@
...
+```
 
-* Conda environment: [RLLib-PhaseOrder/rllib_env.yml](RLLib-PhaseOrder/rllib_env.yml)
-* Build [LLVM-10](https://github.com/llvm/llvm-project/tree/release/10.x)
-    * LLVM can be configured and built with the instructions on this [page](https://llvm.org/docs/CMake.html)
-    * LLVM-10 sources with our custom ODG sub-sequences implemented can be found in [llvm-project-10](llvm-project-10)
-    * Use the following flags to build for AArch64: -DCMAKE_CROSSCOMPILING=True -DLLVM_DEFAULT_TARGET_TRIPLE=aarch64-linux-gnueabihf -DLLVM_TARGET_ARCH=AArch64 -DLLVM_TARGETS_TO_BUILD=AArch64 -DLLVM_ENABLE_PIC=False
-* [IR2Vec](https://github.com/IITH-Compilers/IR2Vec) binary and seed embedding to be used
-* LLVM IR files for training or inference
+#### A small hack to prevent the conda environtments from clashing (To Be removed)
```
Comment (svkeerthy) [inline README.md]: Please recheck if this is needed anymore.

File: README.md (hunk 4 above)
Comment (mvvsmk) [reply README.md:66]: need clarification

File: README.md (hunk 5 above)
Comment (mvvsmk) [reply README.md]: @svkeerthy This was supposed to be one mlbrigde.yml a common virtual env for all envs.

File: README.md [hunk 18]
```diff
@@ -167,11 +167,11 @@ This section will contain information about all the ML driven optimizations. Her
 We propose a Reinforcement Learning (RL) approach for loop distribution, optimizing for both vectorization and locality. Using SCC Dependence Graphs (SDGs), our RL model learns loop distribution order through topological walks. The reward is based on instruction cost and cache misses. We introduce a strategy to expand the training set by generating new loops. This method aims to enhance loop parallelization and improve overall code performance.
 
 This is described in the paper [here](https://ieeexplore.ieee.org/abstract/document/10026979) .
-Please see [here](https://compilers.cse.iith.ac.in/publications/rl_loop_distribution/) for more details.
+Please see [here](https://compilers.cse.iith.ac.in/publications/rl_loop_distribution/) for more details. [Website link](https://compilers.cse.iith.ac.in/publications/rl_loop_distribution/)
```
Comment (svkeerthy) [inline README.md]: Both are same links?

File: README.md [hunk 19]
```diff
@@ -192,9 +192,9 @@ Implimentaion here : [Model Training](./model/LoopDistribution/src/Readme.md) ,
 `RL4ReAl` is a retargetable Reinforcement Learning (RL) approach for solving the REgister ALlocation (REAL) problem on diverse architectures.
 
 This is described in the paper [here](https://dl.acm.org/doi/abs/10.1145/3578360.3580273).
-Please see [here](https://compilers.cse.iith.ac.in/publications/rl4real/) for more details.
+Please see [here](https://compilers.cse.iith.ac.in/publications/rl4real/) for more details. [Website link](https://compilers.cse.iith.ac.in/publications/rl4real/)
```
Comment (svkeerthy) [inline README.md]: Please recheck

File: README.md [hunk 20]
```diff
@@ -192,9 +192,9 @@ Implimentaion here : [Model Training](./model/LoopDistribution/src/Readme.md) ,
...
 
->RL4ReAl: Reinforcement Learning for Register Allocation : S. VenkataKeerthy, Siddharth Jain, Anilava Kundu, Rohit Aggarwal, Albert Cohen, Ramakrishna Upadrasta LLVM-HPC, 2022.
+>RL4ReAl: Reinforcement Learning for Register Allocation : S. VenkataKeerthy, Siddharth Jain, Anilava Kundu, Rohit Aggarwal, Albert Cohen, Ramakrishna Upadrasta CC 2023
```
Comment (svkeerthy) [inline README.md]: Please add ','

**Review by svkeerthy — CHANGES_REQUESTED** [review]

**Review by svkeerthy — APPROVED** [review]
